File: algo/search/js/jump.js
const jump = (nums, key) => {
    const len = nums.length;
    const JUMP_VALUE = Math.trunc(Math.sqrt(len));
    
    let i = 0;
    let j = JUMP_VALUE;

    while (true) {
        if (j > nums.length) {
            j = nums.length;
            break;
        }

        if (key < nums[j]) {
            break;
        }

        i = JUMP_VALUE;
        j += JUMP_VALUE;
    }

    while (i < j) {
        if (nums[i] == key) {
            return i;
        }
        i++;
    }

    return -1;
}

const jump2 = (nums, key) => {
    const len = nums.length;
    const JUMP_VALUE = Math.trunc(Math.sqrt(len));

    let i = 0;
    let step = JUMP_VALUE;

    while (nums[Math.min(step, len) - 1] < key) {
        i = step;
        step += JUMP_VALUE;
        if (i >= len) {
            return -1;
        }
    }

    while (nums[i] < key) {
        i++;
        if (i == Math.min(step, len)) {
            return -1;
        }
    }

    if (nums[i] == key) {
        return i;
    }

    return -1;
}

File: algo/search/java/Jump.java
public class Jump {
    public static void main(String[] args) {
        int key = Integer.parseInt(args[0]);
        System.out.println(Binary.binary(new int []{ 1, 2, 3, 4, 6, 7 }, key));
    }

    public static int jump(int[] nums, int key) {
        final int LEN = nums.length;
        final int JUMP_VALUE = (int ) Math.sqrt(LEN);

        int i = 0;
        int step = JUMP_VALUE;

        while (nums[Math.min(step, LEN)] < key) {
            i = step;
            step += JUMP_VALUE;
            if (i >= LEN) {
                return -1;
            }
        }

        while (nums[i] < key) {
            i++;
            if (i == Math.min(step, LEN)) {
                return -1;
            }
        }
        
        if (nums[i] == key) {
            return i;
        }

        return -1;
    }
}